x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcArray.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.sql.jdbc;
+
+import java.sql.Array;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.sql.jdbc.TypeUtils.baseType;
+
+public class JdbcArray implements Array {
+
+    private final EsType type;
+    private final List<?> values;
+
+    public JdbcArray(EsType type, List<?> values) {
+        this.type = type;
+        this.values = values;
+    }
+    @Override
+    public String getBaseTypeName() throws SQLException {
+        return baseType(type).getName();
+    }
+
+    @Override
+    public int getBaseType() throws SQLException {
+        return baseType(type).getVendorTypeNumber();
+    }
+
+    @Override
+    public Object getArray() throws SQLException {
+        return values.toArray();
+    }
+
+    @Override
+    public Object getArray(Map<String, Class<?>> map) throws SQLException {
+        if (map == null || map.isEmpty()) {
+            return getArray();
+        }
+        throw new SQLFeatureNotSupportedException("getArray with non-empty Map not supported");
+    }
+
+    @Override
+    public Object getArray(long index, int count) throws SQLException {
+        if (index < 0 || index > Integer.MAX_VALUE) {
+            throw new SQLException("Index value [" + index + "] out of range [0, " + Integer.MAX_VALUE + "]");
+        }
+        int index0 = (int) index - 1; // 0-based index
+        if (count < 0) {
+            throw new SQLException("Illegal negative count [" + count + "]");
+        }
+        int available = index0 < values.size() ? values.size() - index0 : 0;
```

Comment:
> From what I can see index can be 0.

It can not, it is [1-based index](https://docs.oracle.com/en/java/javase/15/docs/api/java.sql/java/sql/Array.html#getArray(long,int)).
The out of range exception you exemplified [below](https://github.com/elastic/elasticsearch/pull/68966#discussion_r576726124) was triggered due to an incorrect initial check, `index < 0`, which is now corrected to `< 1` (I also assumed it's 0-based... as an array, so wrote the check, then confronted with the docs, but missed to fix the check :-/).